YALB-258: Add code analysis tools and pre-commit hook

dens0ne opened the pull request:
## [YALB-258](https://yaleits.atlassian.net/browse/YALB-258)

### Description of work
- Adds analysis tools and pre-commit validators found at https://github.com/yalesites-org/eslint-config-and-other-formatting

### Functional testing steps:
- [x] `git checkout` this branch
- [x] Run `npm install` (I am not sure if you need to run anything else particularly)
- [x] Try making a small change to a file and `git commit` it with a message that is not valid in terms of husky standards, and confirm it won't let you commit (I vouch this worked for me locally)

nJim commented on package.json:
```diff
@@ -0,0 +1,28 @@
+    "prepare": "husky install",
+    "prettier": "prettier components --ignore-unknown --list-different",
+    "husky:commit-msg": "commitlint --edit $1",
+    "lint:styles": "stylelint 'components/**/*.scss'",
```
This works well for Drupal themes where we keep CSS in this strucutre, but does not make sense for this profile. There will not be a components directory. Instead, we need to check modules, themes, or whatever we have in this project. I'm thinking this should work:
`"lint:styles": "stylelint **/*.css",`

nJim replied:
Also we should see if there is a pattern for scss and css. I think curly braces can be used to define a pattern? I'm not sure. It's more likely that the yalesites_profile will have css and not scss, but never say never.

dens0ne replied:
> I think curly braces can be used to define a pattern?

I think so, and I think we already have scss coverage already at https://github.com/yalesites-org/yalesites_profile/pull/4/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R24, so I don't think we have to add a pattern for both `css` and `scss`, assuming that I understand things correctly. 

What you are stating makes sense and made me start looking at this with a different perspective. I'll be making another commit soon.

nJim reviewed (changes requested):
Great first steps. I threw a lot of new stuff at you. Even if you used these projects before, this is your first time figuring out the Four Kitchens way of doing it. So good work on everything currently in this PR. Some changes are requested.

#1. Let's add an .nvmrc file to this because newer versions of npm are required for it to install correctly. I know, because my machine was using an old version. You can copy this file and add it to the root of the repo. https://github.com/yalesites-org/component-library-twig/blob/develop/.nvmrc

#2. I added some comments in the PR about the `component` directory not making sense for this project. I suggested some alternative commands. I'm unsure if this easily applues to the lint-stagged stuff too.

#3. Should we track the package-lock.json file? I'm on the fence. Since all sites are supposed to use this common upstream, then committing the lock file makes sense. But I know we've been intentionally skipping the composer lock file. Maybe we should discuss this in the future. No harm done either way.

#4. We are missing the PHP linting and beautifier work to make this static analysis complete. PHP code tools are currently run using PHP Code Sniffer (phpcs and phpcbf.) We typically install this using the drupal coder module so that the Drupal standards are included. Two other dependencies ensure the correct standards are applied. See our example repo in the composer.json for direction.  https://github.com/fourkitchens/linting-and-formatting-standards

#5. Can we add a section to the README.md file about the conventional commits? We can keep it simple. Just a sentence that is project uses conventional commits and then a link going here: https://www.conventionalcommits.org/en/v1.0.0/#summary

dens0ne commented on package.json:
```diff
@@ -0,0 +1,28 @@
+    "prettier": "prettier components --ignore-unknown --list-different",
+    "husky:commit-msg": "commitlint --edit $1",
+    "lint:styles": "stylelint 'components/**/*.scss'",
+    "lint:js": "eslint components",
```
We can double-check what works better, but given documentation at https://github.com/yalesites-org/eslint-config-and-other-formatting states that we should replace `components` with the **path**, I think I'll go with `$INIT_CWD` (see https://docs.npmjs.com/cli/v8/commands/npm-run-script) for now instead... since we want to make sure that it is pointing at the root of the profile. Let me know if you think otherwise.

dens0ne replied:
I take this back, looking at https://github.com/fourkitchens/linting-and-formatting-standards/blob/main/package.json#L13 made me realize it's valide to simply use `.` for current dir, I think that should do it.

dens0ne commented on package.json:
```diff
@@ -0,0 +1,28 @@
+    "@yalesites-org/eslint-config-and-other-formatting": "^1.4.0",
+    "husky": "^7.0.4"
+  },
+  "lint-staged": {
```
@ModulesUnraveled, do you think there is value in adding this `lint-staged` portion to the husky `pre-commit` process (if that's even what's happening...). I think there could be some value but we would have to narrow down the paths to make sure we only account for custom code, but would prefer to get your opinion before thinking about anything esle.

dens0ne commented:
> Let's add an .nvmrc file to this because newer versions of npm are required for it to install correctly. I know, because my machine was using an old version. You can copy this file and add it to the root of the repo. https://github.com/yalesites-org/component-library-twig/blob/develop/.nvmrc

Done.

> I added some comments in the PR about the component directory not making sense for this project. I suggested some alternative commands. I'm unsure if this easily applues to the lint-stagged stuff too.

I accounted for as much as I could, and we should keep discussing how to get our optimal config. Feel free to review the latest changes.

> Should we track the package-lock.json file? I'm on the fence. Since all sites are supposed to use this common upstream, then committing the lock file makes sense. But I know we've been intentionally skipping the composer lock file. Maybe we should discuss this in the future. No harm done either way.

I think we kinda have to skip the `composer.lock` file because (the way I see it) everything within the profile ultimately gets registered in the top level (upstream/base) `composer.lock` file. However, the `package-lock.json` file is a bit different in the sense that you actually run the `npm install` and subsequent commands from within the profile. Thoughts?

> We are missing the PHP linting and beautifier work to make this static analysis complete. PHP code tools are currently run using PHP Code Sniffer (phpcs and phpcbf.) We typically install this using the drupal coder module so that the Drupal standards are included. Two other dependencies ensure the correct standards are applied. See our example repo in the composer.json for direction. https://github.com/fourkitchens/linting-and-formatting-standards

I added the config file and package.json reference, however, in order to get this working, we need to `composer require` phpcs somewhere, and I don't think the installation profile is the right place. We probably should do that in the upstream/base and alter the path to where npm can find the bin. Thoughts?

> Can we add a section to the README.md file about the conventional commits? We can keep it simple. Just a sentence that is project uses conventional commits and then a link going here: https://www.conventionalcommits.org/en/v1.0.0/#summary

Done.

nJim reviewed (approved):
The basic functionality is all here. We have some things to think about for the future. Most noteably, figuring out what need to be in the project vs the profile vs both. Let's work on this organically over time. I have let this PR sit for too long. It's important for us to get our devops setup. So let's ship it and incrementally improve. I will merge in and maybe create another round of changes as I start to word on some related documentation.
